### ISSUE TITLE
Slow first render (generateAndInjectStyles)

### ISSUE THREAD

**Issue by robertontiu** [styled-components/styled-components#2171]
### System:
 - OS: Windows 10
 - CPU: x64 Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz
 - Memory: 1.13 GB / 7.86 GB
### Binaries:
 - Yarn: 1.7.0 - ~\AppData\Roaming\npm\yarn.CMD
 - npm: 6.0.0 - C:\Program Files\nodejs\npm.CMD
 - Watchman: 4.9.4 - C:\Tools\watchman\watchman.EXE
### npmPackages
- babel-plugin-styled-components: 1.7.1
- styled-components: 4.0.0

## Reproduction
The project consists of multiple React root nodes that get rendered server side then hydrated on the client side. For each of the root node, a style tag and markup is outputed on the server side.

After the page is loaded and hydrated on the client, the very first interaction with a react component is very slow (~1s). Further analysis indicates that the function `generateAndInjectStyles` is the one causing the issue. Here is a brief set of results from the performance analysis:

- `generateAndInjectStyles` (@styled-components.browser.esm.js:2025): 805.98ms total time - 0 self time
- `generateAndInjectStyles` (@styled-components.browser.esm.js:1443): 805.98ms total time - 0.29ms self time
- `inject` (@styled-components.browser.esm.js:1080) - 795.61ms total time - 0.38ms self time
- `insertRules` (@styled-components.browser.esm.js:812) - 795.23ms total time - 0ms self time

## Expected Behavior

Either have this run more smoothly or have an option to run the style injection async on DOM ready (or manually).

**Comment by mxstbr** [comment]
IIRC on re-hydration styled-components doesn't inject any styles unless you have checksum mismatches and the styles have changed between server and client, which would mean you need to fix your checksum mismatches!

I'm not 100% sure on that though, pinging @probablyup the resident streaming wizard.

**Comment by robertontiu** [comment]
@mxstbr The problem is actually that the style injection doesn't happen on hydration but rather on first interaction with any react component. If it would happen on hydration (async) it would mitigate the laggish behavior of the first interaction. Regarding checksums, I'm not entirely sure which ones are those (are they `data-styled="<...>"`?)

**Comment by mxstbr** [comment]
I referred React checksum warnings which you see in dev mode, something like this:

```JS
warning.js?3a8d75f:36 Warning: React attempted to reuse markup in a container but the checksum was invalid. This generally means that you are using server rendering and the markup generated on the server was not what the client was expecting. React injected new markup to compensate which works but you have lost many of the benefits of server rendering. Instead, figure out why the markup being generated is different on the client or server:
 (client) "3"><div class="sc-bdVaJa iWyhhR" data-r
 (server) "3"><div class="sc-bwzfXH eJBtGO" data-r
```

> The problem is actually that the style injection doesn't happen on hydration but rather on first interaction with any react component

So, you're saying all the styles get injected all over again the first time a user interacts with the site?

**Comment by robertontiu** [comment]
@mxstbr We don't have any checksum warnings in the console output. 

To sum it up:
- Our website renders components into multiple root nodes and because of that, each of the root components has, next to it (within the body) a style tag generated by styled-component's SSR. 
- On the client side, after hydration and page load, the DOM still contains the style tags next to the markup of the components.
- After the first user interaction, styled-components appears to run `generateAndInjectStyles` which removes those SSR generated style tags and injects a couple of new style tags into the head. (this takes roughly 1 second)
- Any further user interaction works smoothly

**Comment by robertontiu** [comment]
@mxstbr Also, if this is the intented behavior, we would very much like an option to manually trigger this out of the user interaction so we can mitigate the 1 second lag to the page load rather than UX.

**Comment by probablyup** [comment]
> The project consists of multiple React root nodes that get rendered server side then hydrated on the client side. For each of the root node, a style tag and markup is outputted on the server side.

I think this might be the key issue. s-c's SSR implementation wasn't really designed to accommodate multiple React roots and when one loads without the other initially I bet it's clobbering the components that haven't been mounted yet. One more thing to consider for the stylesheet refactor @kitten

**Comment by robertontiu** [comment]
@probablyup Thank you. Are there any temporary workarounds this? (maybe a way to trigger this in the background)

**Comment by probablyup** [comment]
Not at present, sorry. It's all pretty automatic by design.

**Comment by robertontiu** [comment]
@probablyup I did some extra profiling on the matter.

First of all, having one root node doesn't fix the issue. The function (`generateAndInjectStyles`) still gets called on first re-render and takes about 1s. Further calls to this function are very fast. I assume styled-components attempts to do this for all the React tree the first time the reason for it being so slow.

Secondly, I have actually found a temporary workaround:

```ts
import * as React from 'react'
import styled, {css} from 'styled-components'

interface FixLagState {
	ready: boolean
}

const Wrapper = styled.div`
	display: none;
	${(props: FixLagState) => props.ready && css`
		opacity: 0;
	`} 
`

export default class FixLag extends React.Component<{}, FixLagState> {
	state = {
		ready: false
	}

	render() {
		return <Wrapper {...this.state}/>
	}

	componentDidMount() {
		this.setState({ready: true})
	}
}
```

Apparently having this component rendered triggers styled-components in the backround temporarily fixing our issue. It also appears the function takes a lot less since it only has to deal with the styles of this component (being rendered first in the DOM).
